### ISSUE TITLE
reorder collision for different types of sort order

### ISSUE THREAD

**Issue by ComradePashka** [paulzi/yii2-sortable#1]
When you call reorder() function first time with default value for $middle=true so you got sorting with zero in-the-middle, but than later if you try to change sort order type, there could be collision in transaction for some nodes (in case they have same parent_id and overlaps with sortAttribute already set up, integrity constraint for index parent_sort)
At first I thought it is not real issue, because nobody will use reorder() function too often and just have to make unpleasant manual fix to possible dangerous sortAttribute (for example change 100 => 101, 200 => 201, and so on) and then run reorder() again. only for big enough data structure it will be hard as bloody hell to fix them all.
But later I start testing node order function and realized that anyway there is some strange behavior, probably I just simply do not understand how to use it.
I use AL with default sort behavior so data model has field 'sort' and index 'parent_sort', I also had few fields in the model to use MP with AutoTreeTrait before, but now I am using only AL.
so I've got such chain of log messages:
![image](https://cloud.githubusercontent.com/assets/15798005/12093870/79d3bda8-b316-11e5-9aab-f9f99e11c2a7.png)
when I have such data structure: 
![image](https://cloud.githubusercontent.com/assets/15798005/12093945/e7539484-b316-11e5-8937-b9ab3c046c54.png)
and trying to move node id=16 up to one position (change with node id=21) using model method code:

```
            if ($prevNode = $this->prev) {
                $prevNode->insertAfter($this)->save();
            }
```

I run other tests, I was trying to call reorder after any node order changing, and sometimes it works, but sometimes reorder returns zero or have same PDO exception. it looks like reorder() works in different manner if called for different nodes, so I am lost now.

**Comment by paulzi** [comment]
Apparently you have a unique index `parent_sort`. Method `reorder()` changes the sorting elements are not massively, and one by one, so during the operation you may experience the intersection of the sort attribute, because of what the database gives an error unique index. I do not recommend an index on the sort field unique.
Судя по всему у вас индекс `parent_sort` уникальный. Метод переупорядочивания изменяет сортировку узлов не массово, а по одному, поэтому во время работы у вас могут возникать пересечения поля сортировки, из-за чего база данных выдаёт ошибку уникальности индекса. Я не рекомендую делать индекс по полю сортировки уникальным.

**Comment by ComradePashka** [comment]
yep, exactly!
yesterday I double checked everything and find out that if for example you have only two nodes with same parent_id and sort 0 and 100 correspondingly, then after changing nodes and sort order I've got -1 and 0 sort values (but not 100 and 0 as I thought it should be) but at least without any errors in this case. and when I tried to call reorder() or change order with same methods (insertAfter/insertBefore) one more time I've got integrity constraint violation error.
so yeah! I changed my index from unique to simple index and now everything works fine! thank you a lot!
